lib/Dialect/Substrait/IR/Substrait.cpp

```diff
@@ -33,13 +34,96 @@ void SubstraitDialect::initialize() {
       >();
 }
 
+//===----------------------------------------------------------------------===//
+// Substrait interfaces
+//===----------------------------------------------------------------------===//
+
+#include "structured/Dialect/Substrait/IR/SubstraitOpInterfaces.cpp.inc"
+#include "structured/Dialect/Substrait/IR/SubstraitTypeInterfaces.cpp.inc"
+
 //===----------------------------------------------------------------------===//
 // Substrait operations
 //===----------------------------------------------------------------------===//
 
 #define GET_OP_CLASSES
 #include "structured/Dialect/Substrait/IR/SubstraitOps.cpp.inc"
 
+namespace mlir {
+namespace substrait {
+
+/// Verifies that the provided field names match the provided field types. While
+/// the field types are potentially nested, the names are given in a single,
+/// flat list and correspond to the field types in depth first order (where each
+/// nested tuple-typed field has a name and its nested field have names on their
+/// own). Furthermore, the names on each nesting level need to be unique. For
+/// details, see
+/// https://substrait.io/tutorial/sql_to_substrait/#types-and-schemas.
+FailureOr<int> verifyNamedStruct(Location loc,
+                                 llvm::ArrayRef<Attribute> fieldNames,
+                                 TypeRange fieldTypes) {
+  int numConsumedNames = 0;
+  llvm::SmallSet<llvm::StringRef, 8> currentLevelNames;
+  for (Type type : fieldTypes) {
+    // Check name of current field.
+    if (numConsumedNames >= static_cast<int>(fieldNames.size()))
+      return emitError(loc, "not enough field names provided");
+    auto currentName = llvm::cast<StringAttr>(fieldNames[numConsumedNames]);
+    if (!currentLevelNames.insert(currentName).second)
+      return emitError(loc, llvm::Twine("duplicte field name: '") +
```

Comment:
duplicate (should this then be an DictionaryAttr instead?)

Reply:
Typo fixed -- thanks!

Not directly: this models a nested named struct where the unnamed types are given in a nested form and the names in a flat form. Now, the names only need to be unique *per nesting level*, so I can't simply model this type with a dictionary.

I have played with the thought to introduce a type for named structs where each nested level would be a dictionary indeed. But I think the low-tech solution, which doesn't need a custom type, is preferable for now.

Reply:
SG (you can do a dictionaryattr with entries that are dictionaryattr without introducing anything new, but I'm also fine starting here and evaluating)

Reply:
I have thought about the `DictionaryAttr` idea a bit: I think it would be possible to store all types and their names in a dict attr, indeed, but that would be an *attribute*. I still also need the type, which would be redundant. I could of course verify that the redundancy is at least consistent, or even remove it from the assembly with some custom parser/printer but all that sounds more complicated that I'd like this interim solution to be...

Reply:
I did add a TODO about a `NamedStruct` type, though.